src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -0,0 +1,115 @@
+from sentry.utils.outcomes import Outcome, track_outcome
+
+
+def get_metrics_billing_consumer(
+    topic: str,
+    group_id: str,
+    auto_offset_reset: str,
+    **options,
+) -> StreamProcessor[KafkaPayload]:
+    # TODO: support force_topic and force_cluster
+    cluster_name = settings.KAFKA_TOPICS[topic]["cluster"]
+
+    return StreamProcessor(
+        consumer=KafkaConsumer(
+            get_kafka_consumer_cluster_options(
+                cluster_name=cluster_name,
+                override_params={
+                    "enable.auto.commit": False,
+                    "enable.auto.offset.store": False,
+                    "auto.offset.reset": auto_offset_reset,
+                    "group.id": group_id,
+                },
+            )
+        ),
+        topic=Topic(topic),
+        processor_factory=BillingMetricsConsumerStrategyFactory(),
+    )
+
+
+class BillingMetricsConsumerStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    def create_with_partitions(
+        self,
+        commit: Callable[[Mapping[Partition, Position]], None],
+        partitions: Mapping[Partition, int],
+    ) -> ProcessingStrategy[KafkaPayload]:
+        return BillingTxCountMetricConsumerStrategy(commit)
+
+
+class BillingTxCountMetricConsumerStrategy(ProcessingStrategy[KafkaPayload]):
+    """A metrics consumer that generates a billing outcome for each processed
+    transaction, processing a bucket at a time. The transaction count is
+    computed from the amount of values from `d:transactions/duration@millisecond`
+    buckets.
+    """
+
+    def __init__(self, commit: Callable[[Mapping[Partition, Position]], None]) -> None:
+        self.counter_metric_id = TRANSACTION_METRICS_NAMES["d:transactions/duration@millisecond"]
+        self.__commit = commit
+        self.__ready_to_commit: Mapping[Partition, Position] = {}
+        self.__closed = False
+
+    def poll(self) -> None:
+        pass
+
+    def terminate(self) -> None:
+        self.close()
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        assert not self.__closed
+
+        payload = self._get_payload(message)
+        num_processed_transactions = self._count_processed_transactions(payload)
+        self._produce_billing_outcomes(payload, num_processed_transactions)
+        self._mark_commit_ready(message)
+
+    def _get_payload(self, message: Message[KafkaPayload]) -> Dict:
+        return json.loads(message.payload.value.decode("utf-8"), use_rapid_json=True)
+
+    def _count_processed_transactions(self, bucket_payload: Dict) -> int:
+        if bucket_payload["metric_id"] != self.counter_metric_id:
+            return 0
+        return len(bucket_payload["value"])
+
+    def _produce_billing_outcomes(self, payload: Dict, amount: int) -> None:
+        if amount < 1:
+            return
+
+        track_outcome(
```

Comment:
I assume calling this function here, instead of duplicating the logic, isn't a concern in terms of performance and scalability.

Reply:
What does this function do? Keep in mind this is blocking on the submit method (if it's something that can take a while). This way only one message is being processed at a time and the next message won't start until this function is done. Is that ok?

Reply:
It's this function:
https://github.com/getsentry/sentry/blob/87530245448936cea77125c521ece47ee51f9d78/src/sentry/utils/outcomes.py#L39
The function is not heavy computation-wise, so I discarded the alternative to perform the computation asynchronously.

Reply:
`track_outcome` calls `KafkaPublisher.publish`, which itself sends a Kafka message asynchronously, so I _guess_ it's OK?

https://github.com/getsentry/sentry/blob/87530245448936cea77125c521ece47ee51f9d78/src/sentry/utils/outcomes.py#L108

https://github.com/getsentry/sentry/blob/58dbd38e3fa0edc67a19dda3182b9c46fdc126ca/src/sentry/utils/pubsub.py#L74-L75

@lynnagara what would be a better place to call this function? In `self.poll()`, maybe?

Reply:
cc @onewland, this should be cheap.

Reply:
Looking at `track_outcome`, it seems that function uses the legacy publisher to produce messages. That producer does not guarantee delivery as it does not wait for the callback. So you can loose outcomes, specifically in case of errors or when the consumer is shut down.
Is that acceptable ? 
If not please use the [arroyo producer](https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/consumer.py#L676) instead (that returns a future) so you can wait for the future result periodically and commit only when you have confirmation the message has been produced to Kafka.

Reply:
> So you can loose outcomes, specifically in case of errors or when the consumer is shut down.
Is that acceptable ?

From my current understanding, it is acceptable, but I will let @jan-auer make the final call.

Reply:
Let's go with this implementation to unblock other work, and move this to the arroyo producer in an immediate follow-up. We definitely want delivery guarantees.